server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitorTests.java

```diff
@@ -365,4 +373,152 @@ protected void updateIndicesReadOnly(Set<String> indicesToUpdate, ActionListener
+            }
+        };
+
+        final ImmutableOpenMap.Builder<String, DiskUsage> allDisksOkBuilder;
+        allDisksOkBuilder = ImmutableOpenMap.builder();
+        allDisksOkBuilder.put("node1", new DiskUsage("node1", "node1", "/foo/bar", 100, between(15, 100)));
+        final ImmutableOpenMap<String, DiskUsage> allDisksOk = allDisksOkBuilder.build();
+
+        final ImmutableOpenMap.Builder<String, DiskUsage> aboveLowWatermarkBuilder = ImmutableOpenMap.builder();
+        aboveLowWatermarkBuilder.put("node1", new DiskUsage("node1", "node1", "/foo/bar", 100, between(10, 14)));
+        final ImmutableOpenMap<String, DiskUsage> aboveLowWatermark = aboveLowWatermarkBuilder.build();
+
+        final ImmutableOpenMap.Builder<String, DiskUsage> aboveHighWatermarkBuilder = ImmutableOpenMap.builder();
+        aboveHighWatermarkBuilder.put("node1", new DiskUsage("node1", "node1", "/foo/bar", 100, between(5, 9)));
+        final ImmutableOpenMap<String, DiskUsage> aboveHighWatermark = aboveHighWatermarkBuilder.build();
+
+        final ImmutableOpenMap.Builder<String, DiskUsage> aboveFloodStageWatermarkBuilder = ImmutableOpenMap.builder();
+        aboveFloodStageWatermarkBuilder.put("node1", new DiskUsage("node1", "node1", "/foo/bar", 100, between(0, 4)));
+        final ImmutableOpenMap<String, DiskUsage> aboveFloodStageWatermark = aboveFloodStageWatermarkBuilder.build();
+
+        assertNoLogging(monitor, allDisksOk);
+
+        assertSingleInfoMessage(monitor, aboveLowWatermark,
+            "low disk watermark [85%] exceeded on * replicas will not be assigned to this node");
+
+        assertRepeatedWarningMessages(monitor, aboveHighWatermark,
+            "high disk watermark [90%] exceeded on * shards will be relocated away from this node* " +
+                "the node is expected to continue to exceed the high disk watermark when these relocations are complete");
+
+        assertRepeatedWarningMessages(monitor, aboveFloodStageWatermark,
+            "flood stage disk watermark [95%] exceeded on * all indices on this node will be marked read-only");
+
+        relocatingShardSizeRef.set(-5L);
+        assertSingleInfoMessage(monitor, aboveHighWatermark,
+            "high disk watermark [90%] exceeded on * shards will be relocated away from this node* " +
+                "the node is expected to be below the high disk watermark when these relocations are complete");
+
+        relocatingShardSizeRef.set(0L);
+        assertRepeatedWarningMessages(monitor, aboveHighWatermark,
+            "high disk watermark [90%] exceeded on * shards will be relocated away from this node* " +
+                "the node is expected to continue to exceed the high disk watermark when these relocations are complete");
+
+        assertSingleInfoMessage(monitor, aboveLowWatermark,
+            "high disk watermark [90%] no longer exceeded on * but low disk watermark [85%] is still exceeded");
+
+        assertSingleInfoMessage(monitor, allDisksOk,
```

Comment:
Related to my comment above, would be nice to also assert the message when going directly from above high to below low watermark.

I also think we should check that message comes out even if reroute interval did not pass since last reroute?

Reply:
I added assertions about crossing multiple watermarks in ff041dc, including this one.

I added some assertions about the behaviour if time does not pass but it adds significant complexity to make that fully precise. I think we are ok as long as the message is eventually logged. These things are not synchronised with the allocation process so we could already make a decision based on disk watermarks that isn't immediately reflected in the logs.